thunder/benchmarks/__init__.py

```diff
@@ -9,6 +9,7 @@
 from functools import partial
 from numbers import Number
 from typing import Any
+from pytest_benchmark.fixture import BenchmarkFixture
```

Comment:
I think the purpose of using pytest-benchmark is probably to maintain compatibility with the current benchmarking setup (as it supports calibration and can generate both table-formatted output and JSON files used in GitLab's nightly benchmarks). so this PR is kind of fundamentally based on pytest-benchmark. If we decide to move away from pytest-benchmark, we would need to gather requirements and potentially develop a new tool, I'm not sure @IvanYashchuk .

Reply:
Yan started by using a custom benchmark code and I recommended to use pytest benchmark here. I'd like to understand better your concerns. We have used pytest-benchmark for quite some time in `thunder/benchmarks/targets.py` and it proved useful. nvFuser also uses it: https://github.com/NVIDIA/Fuser/blob/4cbd7a4978f23376f9d6dee2efcf981737880a56/benchmarks/python/core.py#L236 Have you communicated your concerns about pytest-benchmark to the nvFuser team?

> it doesn't know about GPU work and so it can make odd choices with how many times to run things

Can you provide an example where not knowing about GPU hurts?

> it runs things a variable number of times it can be really difficult to compare runs with different options to understand what's better

A variable number of runs is done to provide accurate statistics. This behavior can be overwritten with command line options.

> If you put nvtx around the benchmark(function, args) call, then you gave the range for all the calls, and what you really want is the range for one call. Inserting nvtx into the function you're running is difficult but could be an option.

It's straightforward to put the nvtx markers around the function call if that's what is needed, benchmark(add_nvtx(function), args).

Reply:
Hi @tfogal , thank you so much for the long explanation, this PR currently benchmarks each dynamo subgraph (first segmented by dynamo, then split by the thunder splitter, but I think it could be just one subgraph for most litgpt models), on each subgraph(fx GraphModule) we do the compilation using 3 types of backends each(eager, torch.compile, thunder), then the compiled function is passed to the pybenchmark(no nvtx markers), and then the pybench outputs the statistics on each of the subgraphs of 3 backends for comparison, an example could be found [here](https://github.com/Lightning-AI/lightning-thunder/pull/1066#issuecomment-2317778178) 
IIUC, your main concern is that there's no nvtx marker, marker can be easily added in the granularity of the subgraph. And for the overhead of nvtx marker I think it could be small, let me add it and check on some real piece of model and give some data on Monday

Reply:
> my concern is really that pyt-benchmark has made it hard to identify apples-to-apples comparisons.

Can you provide an example?

>And so to wrap that one would naturally do something like:
>```py
> def foo(args):
>   func = thunder.jit(Something())
>   func(*args)
>   torch.cuda.synchronize()
>```
> And then foo is the thing that gets passed to benchmark. 

Of course, if you set up the function-to-be-benchmarked this way you are not measuring the performance of just `func(*args)`. We use a different setup in Thunder "pytest-benchmark"-based benchmarks. `func = thunder.jit(Something())` should be created outside `foo`. `torch.cuda.synchronize()` shouldn't be in `foo`. And then all we're left with inside `foo` is `func(*args)`.

> What one really wants is for pytest-benchmark to make its decisions on the pgm(*args) line alone.

And this is what's happening with pytest-benchmark.

> before line 3103, execute fn(*sample_args) alone (without self.bench) to ensure it's compiled already when line 3103 executes the function.

Let's not do this. It's the benchmarking suite's responsibility to discard first executions that includes compilation as warmup iterations. We ask pytest-benchmark to drop the first iterations here: https://github.com/Lightning-AI/lightning-thunder/blob/952db42d71b2a2019484b05ae74ac828a6946f06/pyproject.toml#L34

> ensure we synchronize (torch.cuda.synchronize()) in whatever function we pass to the BenchmarkFixture. We may need to create a wrapper to do that? Or maybe that's happening already somewhere and it's just hidden from me?

It's hidden in the timer function. We use [`torch.utils.benchmark.timer`](https://github.com/pytorch/pytorch/blob/a063a82c8b166640268d4ebd43b5579c7bf66d01/torch/utils/benchmark/utils/timer.py#L17-L19) and we ask pytest-benchmark to use it here https://github.com/Lightning-AI/lightning-thunder/blob/952db42d71b2a2019484b05ae74ac828a6946f06/pyproject.toml#L33

Reply:
Hi @tfogal 
>before line 3103, execute fn(*sample_args) alone (without self.bench) to ensure it's compiled already when line 3103 executes the function.

I think the pybenchmark warmup can take care of the first time you run fn so that when benchmarking the actually perf it's already compiled and warmed up.
 >ensure we synchronize (torch.cuda.synchronize()) in whatever function we pass to the BenchmarkFixture. We may need to create a wrapper to do that? Or maybe that's happening already somewhere and it's just hidden from me?

 the synchronize is added in https://github.com/Lightning-AI/lightning-thunder/blob/952db42d71b2a2019484b05ae74ac828a6946f06/pyproject.toml#L33-L34 the pytorch timer has the synchronize in it. But if we use it outside the thunder project the command-line option need to be explicitly passed `--benchmark-timer=torch.utils.benchmark.utils.timer.timer` cc: @IvanYashchuk 
 >ensure we get nvtx markers each time the fn is called

 I'll add an option to include the nvtx markers, like the following pattern:
 ```
def nvtx_func(args: list[tensor]):
    nvtx.push_range(...)
    func(*args)
    nvtx.pop_range(...)
benchmark(nvtx_func, args)
 ```